### PR TITLE
#1792 Annotation processor option for default injection strategy

### PR DESCRIPTION
**Pull request by Captain1653**

### CODE REVIEW

File: documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc [hunk 1]
```diff
@@ -293,6 +293,17 @@ Supported values are:
 If a component model is given for a specific mapper via `@Mapper#componentModel()`, the value from the annotation takes precedence.
 |`default`
 
+|`mapstruct.defaultInjectionStrategy`
+| The type of the injection in mapper via parameter `uses`.
+
+Supported values are:
+
+* `field`: dependencies will be injected in fields
+* `constructor`: dependencies will be injected via constructor.
```
Comment (sjaakd) [inline documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc]: perhaps state that a constructor will be generated for this. I was thinking last week though: it should also generate an empty constructor for CDI (does it?), otherwise the object cannot be proxied.

File: documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc [hunk 2]
```diff
@@ -293,6 +293,17 @@ Supported values are:
...
+
+If a injection strategy is given for a specific mapper via `@Mapper#injectionStrategy()`, the value from the annotation takes precedence.
```
Comment (sjaakd) [inline documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc]: what do you mean here?

Comment (Captain1653) [reply documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc]: I wrote it like with component model from our documentation:
"If a component model is given for a specific mapper via @Mapper#componentModel(), the value from the annotation takes precedence." . Was I wrong?

**Comment by Captain1653** [comment]
> Probably add a CDI example.

Yes, you are right. I'll do it, it's my fault.

File: documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc (hunk 1 above)
Comment (Captain1653) [reply documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc]: You're right. componentModel=cdi will generate two constructors: with inject value and default constructor without parameters. I'll add it in documentation.

Comment (Captain1653) [reply documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc]: @sjaakd I'm trying to create tests with CDI, but I can't. Maven (I don't know why) doesn't run my simple 'bad" test. But it compiled it (because I had compile errors with bad imports). I have attached picture. Maybe you know the reason? I use: mvn clean install . Tets 
cdiTest

Comment (sjaakd) [reply documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc]: It is more tricky than that. I tried to the same with `weld-unit` a while ago. Point is: we use a test runner `@RunWith` to trigger annotation processing and we control which classes we want on the class path (we can extend that with `@WithClasses`. Checkout the `testutil`  package for the details.

So, to use another `@RunWith` will not trigger annotation processing. Hence the implementation classes you need will never be generated. 

To solve this, you probably need to use [junit5](https://junit.org/junit5/docs/current/user-guide/#extensions). I believe that @filiphr has tried that. But that's already a tricky job. We probably need the ` @ExtendWith` to do that and use `Order` to first trigger annotation processing with our own runner and then have weld discover the generated implementations.

File: documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc (hunk 2 above)
Comment (sjaakd) [reply documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc]: No you are not wrong.. But perhaps add over what it takes precedence. So: `...takes precedence over the option`. Although it should probably be clear from the context as well.

File: documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc (hunk 1 above)
Comment (Captain1653) [reply documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc]: Ow... I understand. So, about this PR. I think it's better to wait and don't merge it without tests CDI. Should I close?

Comment (sjaakd) [reply documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc]: by the way: I gave up on writing a unit test for that one for now. It is too tricky. But if you want a unit test, you could also go for a fixture. Checkout the way `@Rule` is used in `ConstantsTest` [here](https://github.com/mapstruct/mapstruct/blob/master/processor/src/test/java/org/mapstruct/ap/test/source/constants/ConstantsTest.java).

Comment (sjaakd) [reply documentation/src/main/asciidoc/mapstruct-reference-guide.asciidoc]: > Should I close?

No.. I think your solution is sound. Lets wait for @filiphr's review.

File: documentation/src/main/asciidoc/chapter-2-set-up.asciidoc [hunk 3]
```diff
@@ -254,6 +254,18 @@ Supported values are:
...
 
+|`mapstruct.defaultInjectionStrategy`
+| The type of the injection in mapper via parameter `uses`.
+
+Supported values are:
+
+* `field`: dependencies will be injected in fields
+* `constructor`: will be generated constructor. Dependencies will be injected via constructor.
+
+For `componentModel()=cdi` also will be generated default constructor.
```
Comment (filiphr) [inline documentation/src/main/asciidoc/chapter-2-set-up.asciidoc]: Perhaps reword this to "When cdi `componentModel` a default constructor will also be generated"

File: documentation/src/main/asciidoc/chapter-2-set-up.asciidoc [hunk 4]
```diff
@@ -254,6 +254,18 @@ Supported values are:
 If a component model is given for a specific mapper via `@Mapper#componentModel()`, the value from the annotation takes precedence.
 |`default`
 
+|`mapstruct.defaultInjectionStrategy`
+| The type of the injection in mapper via parameter `uses`.
```
Comment (filiphr) [inline documentation/src/main/asciidoc/chapter-2-set-up.asciidoc]: I think that we should mention somewhere that this is only is only used on annotated based component models such as CDI, Spring and JSR 330. Similar as we don in the Javadoc of `Mapper#injectionStrategy`

File: documentation/src/main/asciidoc/chapter-2-set-up.asciidoc (hunk 3 above)
Comment (Captain1653) [reply documentation/src/main/asciidoc/chapter-2-set-up.asciidoc]: Done

File: documentation/src/main/asciidoc/chapter-2-set-up.asciidoc (hunk 4 above)
Comment (Captain1653) [reply documentation/src/main/asciidoc/chapter-2-set-up.asciidoc]: Done

**Comment by filiphr** [comment]
Thanks for the changes @Captain1653. Just squashed and merged it
